digital_land/update.py

```diff
@@ -66,33 +70,151 @@ def has_collected_resource(log_item):
+        if "end-date" in resource_entry
+        else "",
+    }
+
+    # If empty start-date set by user then no date should be set. Otherwise set user-specified date or use current date
+    if "start-date" in resource_entry:
+        if resource_entry["start-date"]:
+            endpoint_entry["start-date"] = datetime.strptime(
+                resource_entry["start-date"], "%Y-%m-%d"
+            ).date()
+    else:
+        endpoint_entry["start-date"] = date.today().strftime("%Y-%m-%d")
+
+    endpoint_item = Item(endpoint_entry)
+    endpoint_register.add(endpoint_item)
+
+
+def add_new_source(resource_entry, source_register):
+    source_entries = source_register.entries
+    if resource_entry["endpoint"] in source_register.record:
+        for idx in source_register.record[resource_entry["endpoint"]]:
+            if (
+                resource_entry["organisation"]
+                == source_entries[idx].item["organisation"]
+            ):
+                if source_entries[idx].item["end-date"]:
+                    # Entry already exists with an end-date
+                    print(
+                        "WARNING: source end-date {} found for URL {}".format(
+                            source_entries[idx].item["end-date"],
+                            resource_entry["endpoint-url"],
+                        )
+                    )
+                    break
+                else:
+                    # No op if active entry already exists
+                    print(
+                        "Active source entry already exists for organisation {} and URL {}".format(
+                            resource_entry["organisation"],
+                            resource_entry["endpoint-url"],
+                        )
+                    )
+                    return
+
+    source_entry = {
+        "collection": resource_entry.get("pipeline", ""),
+        "pipeline": resource_entry.get("pipeline", ""),
+        "organisation": resource_entry.get("organisation", ""),
+        "endpoint": resource_entry.get("endpoint"),
+        "documentation-url": resource_entry.get("documentation-url", ""),
+        "licence": resource_entry.get("licence", ""),
+        "attribution": resource_entry.get("attribution", ""),
+        "entry-date": date.today().strftime("%Y-%m-%d"),
+        "end-date": datetime.strptime(resource_entry["end-date"], "%Y-%m-%d").date()
+        if "end-date" in resource_entry
+        else "",
+    }
+
+    # If empty start-date set by user then no date should be set. Otherwise set user-specified date or use current date
+    if "start-date" in resource_entry:
+        source_entry["start-date"] = (
+            datetime.strptime(resource_entry["start-date"], "%Y-%m-%d").date()
+            if resource_entry["start-date"]
+            else ""
+        )
+    else:
+        source_entry["start-date"] = date.today().strftime("%Y-%m-%d")
+
+    source_item = Item(source_entry)
```

Comment:
does `Item(source_entry)` _have_ to contain all of the expected keys for a register? or can it be sparsely populated and the register fall back to empty values/defaults?

It feels like a register should allow some empty values, in which case maybe we can avoid explicitly setting all of the optional values for e.g. attribution, licence, end-date etc.

Reply:
So, again, I thought similar - just iterating over `resource_entry` and creating a dict from the existing keys (it would be neater). But I was torn because some entries, e.g. collection, entry-date, end-date have particular handling, so I thought I may as well just make everything explicit. But I can iterate to create the dict and then handle the special cases if that's better?

Reply:
that's fine, explicit is better and I don't think we'd save much. Maybe just do the defaultdict change and that should neaten it up anyway.